lib/puppetserver/ca/local_certificate_authority.rb

```diff
@@ -0,0 +1,177 @@
+require 'puppetserver/ca/host'
+
+require 'openssl'
+
+module Puppetserver
+  module Ca
+    class LocalCertificateAuthority
+
+      # Make the certificate valid as of yesterday, because so many people's
+      # clocks are out of sync.  This gives one more day of validity than people
+      # might expect, but is better than making every person who has a messed up
+      # clock fail, and better than having every cert we generate expire a day
+      # before the user expected it to when they asked for "one year".
+      CERT_VALID_FROM = (Time.now - (60*60*24)).freeze
+
+      SSL_SERVER_CERT = "1.3.6.1.5.5.7.3.1"
+      SSL_CLIENT_CERT = "1.3.6.1.5.5.7.3.2"
+
+      MASTER_EXTENSIONS = [
+        ["basicConstraints", "CA:FALSE", true],
+        ["nsComment", "Puppet Server Internal Certificate", false],
+        ["authorityKeyIdentifier", "keyid:always", false],
+        ["extendedKeyUsage", "#{SSL_SERVER_CERT}, #{SSL_CLIENT_CERT}", true],
+        ["keyUsage", "keyEncipherment, digitalSignature", true],
+        ["subjectKeyIdentifier", "hash", false]
+      ].freeze
+
+      CA_EXTENSIONS = [
+        ["basicConstraints", "CA:TRUE", true],
+        ["keyUsage", "keyCertSign, cRLSign", true],
+        ["subjectKeyIdentifier", "hash", false],
+        ["nsComment", "Puppet Server Internal Certificate", false],
+        ["authorityKeyIdentifier", "keyid:always", false]
+      ].freeze
+
+      def initialize(digest, settings)
+        @digest = digest
+        @host = Host.new(digest)
+        @settings = settings
+      end
+
+      def valid_until
+        Time.now + @settings[:ca_ttl]
+      end
+
+      def extension_factory_for(ca, cert = nil)
+        ef = OpenSSL::X509::ExtensionFactory.new
+        ef.issuer_certificate  = ca
+        ef.subject_certificate = cert if cert
+
+        ef
+      end
+
+      def inventory_entry(cert)
```

Comment:
I didn't know where to put this method.  We need it in both import and generate actions.  Any better ideas would be welcome.

Reply:
This makes sense. This object creates the content for most of those files.